File: dbxml/examples/java/gettingStarted/simpleQuery.java
//
// See the file LICENSE for redistribution information.
//
// Copyright (c) 2004,2009 Oracle.  All rights reserved.
//
//

package dbxml.gettingStarted;

import java.io.*;
import com.sleepycat.dbxml.*;

class simpleQuery
{

    private static String theContainer = "simpleExampleData.dbxml";
    private static String collection="collection('simpleExampleData.dbxml')";
    private static void usage() {
	String usageMessage = "\nThis program performs various queries against a DBXML container.\n";
	usageMessage += "You should run exampleLoadContainer before running this example.\n";
	usageMessage += "You are only required to pass this command the path location of the database\n";
	usageMessage += "environment that you specified when you loaded the examples data:\n\n";

	usageMessage += "\t-h <dbenv directory>\n";

	usageMessage += "For example:\n";
	usageMessage += "\tjava com.sleepycat.dbxml.examples.gettingStarted.simpleQuery -h examplesEnvironment\n";

	System.out.println(usageMessage);
	System.exit( -1 );
    }

    private static void doQuery( XmlManager mgr, String query ) throws Throwable
    {
	//Perform a single query against the referenced container.
	// No context is used for this query.
	String fullQuery = collection + query;
	System.out.println("Exercising query: '" + fullQuery + "'.");
	System.out.println("Return to continue: ");
	System.in.read();

	//Perform the query

	XmlQueryContext context = mgr.createQueryContext();
	XmlResults results = mgr.query(fullQuery, context, null);
	//Iterate over the results of the query using an XmlValue object
	XmlValue value;
	while ((value = results.next()) != null){
            System.out.println(value.asString());
	}
	System.out.println(results.size() + " results returned for query '" +
			   fullQuery + "'.");
	results.delete();
    }

    //Utility function to clean up objects, exceptions or not,
    // containers and environments must be closed.
    private static void cleanup(myDbEnv env, XmlContainer openedContainer) {
	try {
	    if (openedContainer != null)
		openedContainer.delete();
	    if (env != null)
		env.cleanup();
	} catch (Exception e) {
	    // ignore exceptions on close
	}
    }

    public static void main(String args[])
	throws Throwable {

	File path2DbEnv = null;
	for(int i = 0; i < args.length; ++i) {
            if (args[i].startsWith("-")) {
		switch(args[i].charAt(1)) {
		case 'h':
		    path2DbEnv = new File(args[++i]);
		    break;
		default:
                    usage();
		}
            }
	}

	if (path2DbEnv == null || ! path2DbEnv.isDirectory()) {
            usage();
	}

	myDbEnv env = null;
	XmlContainer openedContainer = null;
	try {
	    env = new myDbEnv(path2DbEnv);
	    XmlManager theMgr = env.getManager();

	    //Open a non-transactional container
	    openedContainer =
		theMgr.openContainer(theContainer);

            //Find all the Vendor documents in the database
            doQuery(theMgr, "/vendor");

            //Find all the vendors that are wholesale shops
            doQuery(theMgr, "/vendor[@type=\"wholesale\"]");

            //Find the product document for "Lemon Grass"
            doQuery(theMgr, "/product/item[.=\"Lemon Grass\"]");

            //Find all the products where the price is less than or equal to 0.11
            doQuery(theMgr, "/product/inventory[number(price)<=0.11]");

            //Find all the vegetables where the price is less than or equal to 0.11
            doQuery(theMgr,
		    "/product[number(inventory/price)<=0.11 and category=\"vegetables\"]");
	} catch (Exception e) {
            System.err.println("Error performing query against " + theContainer);
            System.err.println("   Message: " + e.getMessage());
            throw e;
       }
       finally {
	   cleanup(env, openedContainer);
       }
    } //End main
}

File: dbxml/examples/perl/gettingStarted/simpleQuery.pl

use strict;
use warnings;
use Sleepycat::DbXml 'simple';

sub usage()
{
    warn <<EOM ;
This program performs various XPath queries against a DBXML container. 
You should pre-load the container using loadExamplesData.[sh|cmd] before
running this example. You are only required to pass this command the
path location of the database environment that you specified when you
pre-loaded the examples data:

    -h <dbenv directory>
EOM

    exit( -1 );
}

sub doQuery($$$)
{
    my $theMgr = shift ;
    my $container = shift ;
    my $query = shift ;

    #///////////////////////////////////////////////////////////////////////
    #/////  Performs a simple query (no context) against the   ///////
    #/////  provided container.                                      ///////
    #///////////////////////////////////////////////////////////////////////

    #//// some defensive code eliminated for clarity //

    #  Perform the query. Result type is by default Result Document
    my $fullQuery = "collection('" . $container->getName() . "')$query";    
    eval {
        print "Exercising query '$fullQuery '\n" ;
        print "Return to continue: ";
        <> ;

        my $results = $theMgr->query($fullQuery);
        my $value ;
        while( $results->next($value) ) {
            print "$value\n";
        }

        print $results->size() . " objects returned for expression '" .
              "$fullQuery'\n" ;
	};

    if (my $e = catch std::exception)
    {
        warn "Query $fullQuery failed\n";
        warn $e->what() . "\n";
        exit( -1 );
    }
    elsif ($@)
    {
        warn "Query $fullQuery failed\n";
        warn $@;
        exit( -1 );
    }
}

my $path2DbEnv;
my $theContainer = "simpleExampleData.dbxml";
while (@ARGV && $ARGV[0] =~ /^-/ )
{
    my $opt = shift @ARGV ;

    if ($opt =~ s/-h//)
      { $path2DbEnv = $opt || shift }
    else
      { usage() }
}

usage()
  if ! $path2DbEnv ;

eval {
    # open a container in the db environment
    my $env = new DbEnv(0);
    $env->set_cachesize(0, 64 * 1024 * 1024, 1);
    print "opening $path2DbEnv\n";
    $env->open($path2DbEnv,
                Db::DB_INIT_MPOOL|Db::DB_CREATE|Db::DB_INIT_LOCK|Db::DB_INIT_LOG|Db::DB_INIT_TXN);
    my $theMgr = new XmlManager($env);

    my $containerTxn = $theMgr->createTransaction();
    my $container = $theMgr->openContainer($containerTxn, $theContainer);
    $containerTxn->commit();


    # perform the queries

    # find all the Vendor documents in the database
    doQuery( $theMgr, $container, '/vendor' );

    # find all the vendors that are wholesale shops
    doQuery( $theMgr, $container, '/vendor[@type="wholesale"]');

    # find the product document for "Lemon Grass"
    doQuery( $theMgr, $container, '/product/item[.="Lemon Grass"]');

    # find all the products where the price is less than or equal to 0.11
    doQuery( $theMgr, $container, '/product/inventory[price<=0.11]');

    # find all the vegetables where the price is less than or equal to 0.11
    doQuery( $theMgr, $container, '/product[inventory/price<=0.11 and category="vegetables"]');

    exit 0;
} ;

    if (my $e = catch std::exception)
    {
        warn "Query failed\n";
        warn $e->what() . "\n";
        exit( -1 );
    }
    elsif ($@)
    {
        warn "Query failed\n";
        warn $@;
        exit( -1 );
    }
